### PR TITLE
[config] Clean up toDecryptedPropertyValue

### PR DESCRIPTION
**Pull request by pgavlin**
- Plumb a `context.Context` parameter to pass to the decrypter
- Rephrase conversion over `config.Plaintext` values in order to decouple decryption and conversion

### CODE REVIEW

File: sdk/go/common/resource/config/object.go [hunk 1]
```diff
@@ -524,36 +524,10 @@ func isSecureValue(v any) (bool, string) {
 	return false, ""
 }
 
-func (c object) toDecryptedPropertyValue(decrypter Decrypter) resource.PropertyValue {
-	var prop resource.PropertyValue
-	switch v := c.value.(type) {
-	case bool, int64, float64, string:
-		if c.secure {
-			plaintext, err := decrypter.DecryptValue(context.Background(), v.(string))
-			contract.AssertNoErrorf(err, "failed to decrypt config")
```
Comment (pgavlin) [inline sdk/go/common/resource/config/object.go:533]: For the future: we cannot have asserts like this. It is reasonable for decryption to fail, and we should not crash if it does. We should return errors instead and let callers handle them.

File: sdk/go/common/resource/config/plaintext.go [hunk 2]
```diff
@@ -117,6 +152,8 @@ func (c Plaintext) Encrypt(ctx context.Context, encrypter Encrypter) (Value, err
 // encrypt converts the receiver to an object. All secure strings in the result are encrypted using encrypter.
 func (c Plaintext) encrypt(ctx context.Context, path resource.PropertyPath, encrypter Encrypter) (object, error) {
 	switch v := c.Value().(type) {
+	case nil:
+		return object{}, nil
```
Comment (tgummerer) [inline sdk/go/common/resource/config/plaintext.go:156]: This seems unrelated to the cleanup.  Is this fixing anything in particular, or just defensive because we've seen similar problems in other places?

**Review by tgummerer — APPROVED** [review]
Thanks for the cleanup!

**Review by Frassle — APPROVED** [review]
Looks reasonable

File: sdk/go/common/resource/config/plaintext.go (hunk 2 above)
Comment (tgummerer) [reply sdk/go/common/resource/config/plaintext.go:156]: I just saw https://github.com/pulumi/pulumi/issues/14570, which this will fix :sweat_smile:
